### 异常

异常： 就是程序在运行时出现不正常的情况

异常由来： Java对异常类进行描述和封装。

对于问题的划分：

1. 严重的

   1. Java通过Error类进行描述
   2. 一般不编写针对性的代码进行处理

2. 非严重的问题

   对与非严重的，Java通过Exception类进行描述。

   对于Exception可以使用针对性的处理方式进行处理。



无论Error和expection 都有一些共性内容

Throwable

### 处理代码异常 try catch 

```java
try{}catch(Exception e){}finally{}
```



### throws

```
class Demo{
	int div(int a, int b)throw Exception{ //在功能上通过throws关键字声明了该功能可能会有问题。
		 return a/b;
	}
}
// Exception 也可以自定义一个类 比如throw 自定义类
```

### 多异常

1. 声明异常时，建议声明更为具体的异常。这样处理可以更具体。
2. 对方声明几个异常，就对应有几个catch块。不要定义多余的catch块。
   1. 如果多个catch块中的异常出现继承关系，父类异常catch块放在最下面

3. 建议进行catch处理时，catch中一定要定义具体处理方式。不要定义简单的一句e.printStackTrace();



### throws和throw的区别

1. throws使用在函数上，throw使用在函数内
2. throws后面跟的异常类。可以跟多个，用逗号隔开。throw跟的是对象。



###  RuntimeExpection

Exception 中有一个特殊的子类异常RuntimeExpection 运行时异常。

如果在函数内容抛出该异常，函数上可以不用声明。编译一样通过。

如果在函数上声明了该异常。调用者可以不用进行处理，编译一样通过。

之所以不用在函数声明，是因为不需要让调用者处理。

当该异常发生，希望程序停止，因为在运行时，出现了无法继续运算的情况，希望停止程序后，对代码进行修正。



### 异常练习

```
/**
 * 有一个圆形和长方形。
 * 都可以获取面积，对于面积如果出现非法的数值，视为是获取面积出现了问题
 * 问题通过异常表示。
 *RuntimeException 直接中断程序。 表示不可处理
 * Exception 可处理的
 *
 * 
 */

interface Shape{
    double getArea();

}
class NoValueException extends RuntimeException{
    NoValueException(String message){
        super(message);
    }
}
class Circle{
    private int radius;
    private static final double PI = 3.14;

    Circle(int radius){
        this.radius = radius;
    }
    public void getArea(){
        System.out.println(PI * radius);
    }
}
class Rec implements Shape{
    int x,y;
    Rec(int x ,int y){
        if(x<=0 || y<=0 ){
            throw new NoValueException("出现非法值");
        }
        this.x = x;
        this.y = y;
    }

    @Override
    public double getArea() {
        return x*y;
    }
}

class ExceptionTest1{
    public static void main(String[] args) {
            Rec r = new Rec(3,4);
            r.getArea();
    }
}
```



### 异常总结

异常是对问题的描述，将问题进行对象的封装。

异常体系：

 * Throwable
    * Error
    * Exception
       * RuntimeException

异常体系的特点： 异常体系中的所有类以及建立的对象都具备可抛性

也就是说可以被throw和throws关键字操作

只有异常体系具备这个特点



throw和throws的用法：

1. throw定义在函数内，用于抛出异常对象。

2. throws定义在函数上，用于抛出异常类，可以抛出多个用逗号隔开。

3. 当函数内容有throw抛出异常对象，并未进行try处理。必须要在函数上声明，否则编译失败。

注意： RuntimeException除外也就是说，函数内如果抛出的RuntimeException异常。函数上可以不用声明。

如果函数声明了异常，调用者需要进行处理，处理方法可以throws可以try。



异常有两种：

 	1. 编译时被检测异常
      	1. 该异常在编译时没有处理，那么编译失败。
      	2. 该异常被标示代表着可以被处理
 	2. 运行时异常（编译时不检测）
      	1. 在编译时，不需要处理，编译器不检查.
      	2. 该异常的发生，建议不处理。让程序停止，需要对代码进行修正。



finally

1. finally中定义的通常是，关闭资源代码。
2. 在遇到system.exit(0) 不会执行。



自定义异常

定义类继承exception 或者runtimeException；

1. 为了让该类自定义具备可抛性
2. 让该类具备操作异常的共性方法。

当要定义自定义异常的信息时，可以使用父类已经定义好的功能。

异常信息传递给父类的构造函数。

```java
class MyException extends Exception{
	MyException(String message){
		super(message)
	}
}
```



### 异常的好处

1. 将问题进行封装
2. 将正常流程代码和问题处理代码相分离，方便与阅读。



### 异常的处理原则：

1. 处理方式有两种： try或者throws
2. 调用到抛出异常的功能时，抛出几个，就处理几个。
   1. 一个try对应多个catch的情况
3. 多个catch，父类的catch放到最下面
4. catch内需要定义针对性的处理方式。不要简单的定义输出语句等。
5. 当捕获到的异常，本功能处理不了时，可以继续在catch中抛出
6. 如果该异常处理不了，但并不属于该功能出现的异常。可以将异常转换后，在抛出和该功能相关的异常
7. 或者异常可以处理，当需要将异常产生的和本功能相关的问题提供出去，当调用者知道，并处理，也可以将捕获异常处理后，转换新的异常。